// es-abstract/2019/NumberToRawBytes.js
'use strict';

var GetIntrinsic = require('get-intrinsic');

var $TypeError = GetIntrinsic('%TypeError%');

var hasOwnProperty = require('es-abstract/2019/HasOwnProperty');
var ToInt16 = require('es-abstract/2019/ToInt16');
var ToInt32 = require('es-abstract/2019/ToInt32');
var ToInt8 = require('es-abstract/2019/ToInt8');
var ToUint16 = require('es-abstract/2019/ToUint16');
var ToUint32 = require('es-abstract/2019/ToUint32');
var ToUint8 = require('es-abstract/2019/ToUint8');
var ToUint8Clamp = require('es-abstract/2019/ToUint8Clamp');
var Type = require('es-abstract/2019/Type');

var valueToFloat32Bytes = require('es-abstract/helpers/valueToFloat32Bytes');
var valueToFloat64Bytes = require('es-abstract/helpers/valueToFloat64Bytes');
var integerToNBytes = require('es-abstract/helpers/integerToNBytes');

var keys = require('object-keys');

// https://262.ecma-international.org/8.0/#table-50
var TypeToSizes = {
	__proto__: null,
	Int8: 1,
	Uint8: 1,
	Uint8C: 1,
	Int16: 2,
	Uint16: 2,
	Int32: 4,
	Uint32: 4,
	Float32: 4,
	Float64: 8
};

var TypeToAO = {
	__proto__: null,
	Int8: ToInt8,
	Uint8: ToUint8,
	Uint8C: ToUint8Clamp,
	Int16: ToInt16,
	Uint16: ToUint16,
	Int32: ToInt32,
	Uint32: ToUint32
};

// https://262.ecma-international.org/8.0/#sec-numbertorawbytes

module.exports = function NumberToRawBytes(type, value, isLittleEndian) {
	if (typeof type !== 'string' || !hasOwnProperty(TypeToSizes, type)) {
		throw new $TypeError('Assertion failed: `type` must be a TypedArray element type: ' + keys(TypeToSizes));
	}
	if (Type(value) !== 'Number') {
		throw new $TypeError('Assertion failed: `value` must be a Number');
	}
	if (Type(isLittleEndian) !== 'Boolean') {
		throw new $TypeError('Assertion failed: `isLittleEndian` must be a Boolean');
	}

	if (type === 'Float32') { // step 1
		return valueToFloat32Bytes(value, isLittleEndian);
	} else if (type === 'Float64') { // step 2
		return valueToFloat64Bytes(value, isLittleEndian);
	} // step 3

	var n = TypeToSizes[type]; // step 3.a

	var convOp = TypeToAO[type]; // step 3.b

	var intValue = convOp(value); // step 3.c

	return integerToNBytes(intValue, n, isLittleEndian); // step 3.d, 3.e, 4
};
